src/lib/communication/DY.Lib.Communication.API.Lemmas.fst

```diff
@@ -0,0 +1,312 @@
+module DY.Lib.Communication.API.Lemmas
+
+open Comparse
+open DY.Core
+open DY.Lib.Crypto.PkEncryption.Split
+open DY.Lib.Crypto.Signature.Split
+open DY.Lib.State.PKI
+open DY.Lib.State.PrivateKeys
+open DY.Lib.Event.Typed
+
+open DY.Lib.Communication.API.Predicates
+open DY.Lib.Communication.API
+
+#set-options "--fuel 0 --ifuel 0 --z3cliopt 'smt.qi.eager_threshold=100'"
+
+(**** Confidential Send and Receive Lemmas ****)
+
+val encrypt_message_proof:
+  {|cinvs:crypto_invariants|} ->
+  tr:trace ->
+  sender:principal -> receiver:principal -> payload:bytes ->
+  pk:bytes -> nonce:bytes ->
+  Lemma
+  (requires
+    has_communication_layer_invariants cinvs /\
+    is_knowable_by (join (principal_label sender) (principal_label receiver)) tr payload /\
+    is_secret (principal_label sender) tr nonce /\
+    PkNonce? (get_usage nonce) /\
+    is_encryption_key (PkKey comm_layer_pkenc_tag empty) (principal_label receiver) tr pk /\
+    event_triggered tr sender (CommConfSendMsg sender receiver payload)
+  )
+  (ensures
+    is_publishable tr (encrypt_message sender receiver payload pk nonce)
+  )
+let encrypt_message_proof #cinvs tr sender receiver payload pk nonce =
+  let msg = CM {sender; receiver; payload} in  
+  serialize_wf_lemma communication_message (is_knowable_by (join (principal_label sender) (principal_label receiver)) tr) msg;
+  ()
+
+val send_confidential_proof:
+  {|invs:protocol_invariants|} ->
+  tr:trace ->
+  higher_layer_preds:comm_higher_layer_event_preds ->
+  keys_sess_ids:communication_keys_sess_ids ->
+  sender:principal -> receiver:principal -> payload:bytes ->
+  Lemma
+  (requires
+    trace_invariant tr /\
+    has_pki_invariant invs /\
+    has_communication_layer_invariants invs.crypto_invs /\
+    comm_higher_layer_event_preds_later tr (higher_layer_preds.send_conf sender receiver payload) /\
+    has_communication_layer_event_predicates invs higher_layer_preds /\    
+    is_knowable_by (join (principal_label sender) (principal_label receiver)) tr payload
```

Comment:
```suggestion
    is_knowable_by (join (principal_label sender) (principal_label receiver)) tr payload /\
    comm_higher_layer_event_preds_later tr (higher_layer_preds.send_conf sender receiver payload) /\
    trace_invariant tr /\
    has_pki_invariant invs /\
    has_communication_layer_invariants invs.crypto_invs /\
    has_communication_layer_event_predicates invs higher_layer_preds
```
When a lemma has a lot of preconditions, I find it useful to group them by category (e.g. the `has_..._invariants`) and order them from most specific to this lemma to less specific (e.g. the `has_..._invariants` are everywhere, so put it at the end).
This could be done in other lemmas of this file